core-bundle/src/InsertTag/InsertTagParser.php

```diff
@@ -12,65 +12,625 @@
+use Symfony\Component\HttpKernel\Controller\ControllerReference;
+use Symfony\Component\HttpKernel\Fragment\FragmentHandler;
 use Symfony\Contracts\Service\ResetInterface;
 
+/**
+ * Parse and replace insert tags.
+ *
+ * Formal syntax (EBNF):
+ *
+ *        InsertTag ::= "{{" Name Parameter * Flag * "}}"
+ *             Name ::= [a-z#x80-#xFF] [a-z0-9_#x80-#xFF] *
+ *        Parameter ::= "::" ( KeyValuePair | Value )
+ *             Flag ::= "|" [^{}|] *
+ *     KeyValuePair ::= Key "=" Value
+ *              Key ::= [^{}|=] *
+ *            Value ::= ( [^{}|] | InsertTag ) *
+ */
 class InsertTagParser implements ResetInterface
 {
+    private const TAG_REGEX = /** @lang RegExp */ '
+        (?<it>                 # Named capturing group "it"
+            {{                 # Starts with two opening curly braces
+            [a-z0-9\x80-\xFF]  # The first letter must not be a reserved character of Twig, Mustache or similar template engines (see #805)
+            (?>[^{}]|(?&it))*  # Match any character not curly brace or a nested insert tag
+            }}                 # Ends with two closing curly braces
+        )';
+
+    private const PARAMETER_REGEX = /** @lang RegExp */ '
+        ::                        # Starts with double colon
+        (?:
+            [^{}|:]               # Match any character not curly brace, pipe or colon
+            |:(?!:)               # Or a single colon (not followed by another colon)
+            |'.self::TAG_REGEX.'  # Or an insert tag
+        )*';
+
+    /**
+     * @var array<string, InsertTagSubscription>
+     */
+    private array $subscriptions = [];
+
+    /**
+     * @var array<string, InsertTagSubscription>
+     */
+    private array $blockSubscriptions = [];
+
+    /**
+     * @var array<string, \Closure(InsertTagFlag, InsertTagResult):InsertTagResult>
+     */
+    private array $flagCallbacks = [];
+
+    private readonly string $allowedTagsRegex;
+
     public function __construct(
         private readonly ContaoFramework $framework,
+        private readonly LoggerInterface $logger,
+        private readonly FragmentHandler $fragmentHandler,
+        private readonly RequestStack $requestStack,
         private InsertTags|null $insertTags = null,
+        array $allowedTags = ['*'],
     ) {
+        $this->allowedTagsRegex = '('.implode(
+            '|',
+            array_map(
+                static fn ($allowedTag) => '^'.implode('.+', array_map('preg_quote', explode('*', $allowedTag))).'$',
+                $allowedTags ?: [''],
+            ),
+        ).')';
     }
 
-    public function replace(string $input): string
+    public function addSubscription(InsertTagSubscription $subscription): void
     {
-        return (string) $this->replaceChunked($input);
+        if (1 !== preg_match($this->allowedTagsRegex, $subscription->name)) {
+            return;
+        }
+
+        unset($this->blockSubscriptions[$subscription->name]);
```

Comment:
Any reason to silently move things from a block to non-block subscription (and vice versa) instead of throwing an exception, here?

Reply:
Changed to exceptions in #6195